src/impression.js

```diff
@@ -47,15 +63,59 @@ export function maybeTrackImpression(win) {
     // avoid duplicate tracking.
     win.location.hash = '';
   }
+
   viewer.whenFirstVisible().then(() => {
-    invoke(win, clickUrl);
+    // TODO(@zhouyx) We need a timeout here?
+    // TODO(@zhouyx) need test with a real response.
+    invoke(win, dev().assertString(clickUrl)).then(response => {
+      applyResponse(win, viewer, response);
+      resolveImpression();
+    });
   });
 }
 
+/**
+ * Send the url to ad server and wait for its response
+ * @param {!Window} win
+ * @param {string} clickUrl
+ * @return {!Promise<!JSONType>}
+ */
 function invoke(win, clickUrl) {
-  xhrFor(win).fetchJson(clickUrl, {
+  if (getMode().localDev && !getMode().test) {
+    clickUrl = 'http://localhost:8000/impression-proxy?url=' + clickUrl;
+  }
+  return xhrFor(win).fetchJson(clickUrl, {
     credentials: 'include',
     requireAmpResponseSourceOrigin: true,
   });
-  // TODO(@cramforce): Do something with the result.
+}
+
+/**
+ * parse the response back from ad server
+ * Set for analytics purposes
+ * @param {!Window} win
+ * @param {!Object} response
+ */
+function applyResponse(win, viewer, response) {
+  const adLocation = response['location'];
+  const adTracking = response['tracking'];
+
+  // If there is a tracking_url, need to track it
+  // Otherwise track the location
+  const trackUrl = adTracking || adLocation;
+  if (trackUrl) {
+    new Image().src = trackUrl;
+  }
+
+  // If have we have gclid replace the location href with new location we get.
+  // TODO(@zhouyx) should we replaceState directly w/o checking gclid?
+  const gclid = response['gclid'];
+  if (gclid && adLocation) {
+    if (getMode().localDev && !getMode().test) {
+      win.history.replaceState(null, '', win.location.href +
+          '#gclid=' + gclid);
+      return;
+    }
+    win.history.replaceState(null, '', adLocation);
```

Comment:
I see a lot of usage of `win.history.replaceState`. @dvoytenko if I need to use `HistoryService` how should I use that?